File: src/go/ch01/1.3/l2/demo_number_theory_advanced.go
package main

import "fmt"

func main() {
	// Extended Euclid / inverse
	g, x, y := ExtendedGCD(240, 46)
	fmt.Println("extended_gcd(240,46) ->", g, x, y)
	if inv, err := InvMod(3, 7); err == nil {
		fmt.Println("invmod(3,7) =", inv)
	} else {
		fmt.Println("invmod(3,7) failed:", err)
	}

	// Totient
	if v, _ := PhiU64(10); true {
		fmt.Println("phi(10) =", v)
	}
	if v, _ := PhiU64(36); true {
		fmt.Println("phi(36) =", v)
	}

	// Miller–Rabin
	for _, n := range []uint64{97, 561, 1105, 2147483647} {
		fmt.Printf("is_probable_prime(%d) = %v\n", n, IsProbablePrimeU64(n))
	}

	// CRT: x≡2 (mod 3), x≡3 (mod 5), x≡2 (mod 7) -> 23 mod 105
	x12, m12, _ := CRTPairU64(2, 3, 3, 5)
	res, mod, _ := CRTPairU64(x12, m12, 2, 7)
	fmt.Println("CRT ->", res, "mod", mod)

	// Exercises (optional to print)
	runExercises()
}

File: src/erlang/ch01/1.3/l2/src/demo_number_theory_advanced.erl
-module(demo_number_theory_advanced).
-export([main/0]).

main() ->
    %% Extended Euclid / inverse
    {G, X, Y} = extgcd_inv:extended_gcd(240, 46),
    io:format("extended_gcd(240,46) -> ~p~n", [{G,X,Y}]),
    io:format("invmod(3,7) = ~p~n", [extgcd_inv:invmod(3,7)]),

    %% Totient
    io:format("phi(10) = ~p~n", [totient:phi(10)]),
    io:format("phi(36) = ~p~n", [totient:phi(36)]),

    %% Miller–Rabin
    lists:foreach(
      fun(N) ->
          io:format("is_probable_prime(~p) = ~p~n", [N, miller_rabin:is_probable_prime(N)])
      end,
      [97, 561, 1105, 2147483647]),

    %% CRT: x≡2 (mod 3), x≡3 (mod 5), x≡2 (mod 7) -> 23 mod 105
    {X12, M12} = crt:crt_pair({2,3}, {3,5}),
    {Res, Mod} = crt:crt_pair({X12, M12}, {2,7}),
    io:format("CRT -> ~p mod ~p~n", [Res, Mod]).